src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -717,11 +852,12 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
      * @param includeSensitive
      * @throws Exception
      */
-    public Map<String, Integer> writeResultsFromIndexToStream(final DownloadRequestParams downloadParams,
+    public ConcurrentMap<String, AtomicInteger> writeResultsFromIndexToStream(final DownloadRequestParams downloadParams,
                                                                          OutputStream out,
                                                                          boolean includeSensitive, final DownloadDetailsDTO dd, boolean checkLimit) throws Exception {
+        ExecutorService nextExecutor = getSolrThreadPoolExecutor();
```

Comment:
DownloadControlThread does use getSolrThreadPoolExecutor. 

Online requests:
OccurrenceController ->
DownloadService.writeQueryToStream() -> 
DownloadService.writeQueryToStream() -> 
SearchDAOImpl.writeResultsFromIndexToStream() -> SearchDAOImpl.getSolrThreadPoolExecutor()

Offline requests:
DownloadControlThread.run() ->
DownloadServiceExecutor.submitDownload() ->
DownloadService.DownloadCreatorImpl.createCallable() -> 
DownloadService.writeQueryToStream() -> 
SearchDAOImpl.writeResultsFromIndexToStream() -> SearchDAOImpl.getSolrThreadPoolExecutor()

As the code is now, offline downloads can block solrExecutor from executing online downloads because SearchDAOImpl.writeResultsFromIndexToStream(), which is used by both, will produce 1 to 24 callables for each request to be consumed by solrExecutor.

Using a separate executor pool for online and offline requests would fix this and appeared to be inline with the approach that you have taken. 

My preference would be to create a new executor for each getSolrThreadPoolExecutor request.